File: src/main/java/org/embulk/filter/copy/CopyFilterPlugin.java
package org.embulk.filter.copy;

import org.embulk.config.Config;
import org.embulk.config.ConfigDefault;
import org.embulk.config.ConfigSource;
import org.embulk.config.Task;
import org.embulk.config.TaskSource;
import org.embulk.filter.copy.forward.InForwardService;
import org.embulk.filter.copy.forward.OutForwardEventBuilder;
import org.embulk.filter.copy.forward.OutForwardService;
import org.embulk.filter.copy.plugin.InternalForwardInputPlugin;
import org.embulk.filter.copy.runner.AsyncEmbulkRunnerService;
import org.embulk.filter.copy.runner.EmbulkRunner;
import org.embulk.filter.copy.spi.PageBuilder;
import org.embulk.filter.copy.spi.PageReader;
import org.embulk.filter.copy.spi.StandardColumnVisitor;
import org.embulk.spi.ColumnVisitor;
import org.embulk.spi.Exec;
import org.embulk.spi.FilterPlugin;
import org.embulk.spi.Page;
import org.embulk.spi.PageOutput;
import org.embulk.spi.Schema;
import org.slf4j.Logger;

import java.util.List;

public class CopyFilterPlugin
        implements FilterPlugin
{
    private final static Logger logger = Exec.getLogger(CopyFilterPlugin.class);

    public interface EmbulkConfig
            extends Task
    {
        @Config("exec")
        @ConfigDefault("{}")
        ConfigSource getExecConfig();

        @Config("filters")
        @ConfigDefault("[]")
        List<ConfigSource> getFilterConfig();

        @Config("out")
        ConfigSource getOutputConfig();
    }

    public interface PluginTask
            extends Task, OutForwardService.Task, InForwardService.Task
    {
        @Config("config")
        EmbulkConfig getConfig();
    }

    private EmbulkRunner configure(PluginTask task, Schema schema)
    {
        ConfigSource inputConfig = Exec.newConfigSource();
        inputConfig.set("type", InternalForwardInputPlugin.PLUGIN_NAME);
        inputConfig.set("columns", schema);
        inputConfig.set("message_tag", task.getMessageTag());
        inputConfig.set("shutdown_tag", task.getShutdownTag());
        inputConfig.set("in_forward", task.getInForwardTask());

        return EmbulkRunner.builder()
                .execConfig(task.getConfig().getExecConfig())
                .inputConfig(inputConfig)
                .filterConfig(task.getConfig().getFilterConfig())
                .outputConfig(task.getConfig().getOutputConfig())
                .build();
    }

    private void withEmbulkRun(AsyncEmbulkRunnerService service, Runnable r)
    {
        service.startAsync();
        service.awaitRunning();
        r.run();
        service.awaitTerminated();
    }

    @Override
    public void transaction(ConfigSource config, Schema inputSchema,
            FilterPlugin.Control control)
    {
        PluginTask task = config.loadConfig(PluginTask.class);
        EmbulkRunner embulkRunner = configure(task, inputSchema);
        AsyncEmbulkRunnerService embulkRunnerService = new AsyncEmbulkRunnerService(embulkRunner);

        withEmbulkRun(embulkRunnerService, () -> {
            Schema outputSchema = inputSchema;
            control.run(task.dump(), outputSchema);
            OutForwardService.sendShutdownMessage(task);
        });
    }

    @Override
    public PageOutput open(TaskSource taskSource, final Schema inputSchema,
            final Schema outputSchema, final PageOutput output)
    {
        final PluginTask task = taskSource.loadTask(PluginTask.class);

        return new PageOutput()
        {
            private final PageReader pageReader = new PageReader(inputSchema);
            private final PageBuilder pageBuilder = new PageBuilder(Exec.getBufferAllocator(), outputSchema, output);
            private final StandardColumnVisitor standardColumnVisitor = new StandardColumnVisitor(pageReader, pageBuilder);
            private final OutForwardService outForwardService = new OutForwardService(task);
            private final OutForwardEventBuilder eventBuilder = new OutForwardEventBuilder(outputSchema, outForwardService);
            private final ColumnVisitor outForwardVisitor = new StandardColumnVisitor(pageReader, eventBuilder);

            @Override
            public void add(Page page)
            {
                pageReader.setPage(page);
                while (pageReader.nextRecord()) {
                    outputSchema.visitColumns(outForwardVisitor);
                    eventBuilder.addRecord();

                    outputSchema.visitColumns(standardColumnVisitor);
                    pageBuilder.addRecord();
                }
            }

            @Override
            public void finish()
            {
                outForwardService.finish();
                pageBuilder.finish();
            }

            @Override
            public void close()
            {
                outForwardService.close();
                pageBuilder.close();
            }

        };
    }
}

File: src/main/scala/pro/civitaspo/embulk/filter/copy/CopyFilterPlugin.scala
package pro.civitaspo.embulk.filter.copy

import java.util.{Optional, List => JList}

import org.embulk.config.{
  Config,
  ConfigDefault,
  ConfigException,
  ConfigSource,
  Task,
  TaskSource
}
import org.embulk.spi.{
  DataException,
  Exec,
  FilterPlugin,
  Page,
  PageBuilder,
  PageOutput,
  Schema
}
import org.slf4j.{Logger, LoggerFactory}

import scala.concurrent.{Await, ExecutionContextExecutor, Future}
import scala.concurrent.duration.Duration
import scala.util.{Failure, Success}
import scala.util.chaining._

class CopyFilterPlugin extends FilterPlugin {
  import implicits._

  private val logger: Logger =
    LoggerFactory.getLogger(classOf[CopyFilterPlugin])

  trait PluginTask extends Task {
    @Config("config")
    @ConfigDefault("null")
    def getConfig: Optional[BreakinBulkLoader.Task]

    @Config("copy")
    @ConfigDefault("[]")
    def getCopy: JList[BreakinBulkLoader.Task]
  }

  override def transaction(
      config: ConfigSource,
      inputSchema: Schema,
      control: FilterPlugin.Control
  ): Unit = {
    val task = config.loadConfig(classOf[PluginTask])
    task.getConfig.foreach { cfg =>
      logger.warn("[DEPRECATED] Use \"copy\" option instead.")
      task.getCopy.add(cfg)
    }
    if (task.getCopy.isEmpty)
      throw new ConfigException("Either 'copy' or 'config' option is required.")
    val transactionId: String = Utils.genTransactionId()
    task.getCopy.foreach(_.setTransactionId(transactionId))

    withRunningBulkLoaders(task, inputSchema, transactionId) {
      control.run(task.dump(), inputSchema)
    }
  }

  override def open(
      taskSource: TaskSource,
      inputSchema: Schema,
      outputSchema: Schema,
      output: PageOutput
  ): PageOutput = {
    val task = taskSource.loadTask(classOf[PluginTask])

    val copyOutputs: Seq[PageOutput] = task.getCopy.zipWithIndex
      .map(x => BreakinBulkLoader(x._1, x._2))
      .map(l => Utils.buildCopyPageOutput(l.sendPage))

    new PageOutput {
      private val pageBuilders: Seq[PageBuilder] = (copyOutputs :+ output).map(
        new PageBuilder(Exec.getBufferAllocator, inputSchema, _)
      )
      private val visitor = CopyColumnVisitor(inputSchema, pageBuilders)
      override def add(page: Page): Unit = visitor.visit(page)
      override def finish(): Unit = pageBuilders.foreach(_.finish())
      override def close(): Unit = pageBuilders.foreach(_.close())
    }
  }

  private def withRunningBulkLoaders(
      task: PluginTask,
      schema: Schema,
      transactionId: String
  )(
      runControl: => Unit
  ): Unit = {
    val loaders: Seq[BreakinBulkLoader] = task.getCopy.zipWithIndex
      .map(x => BreakinBulkLoader(x._1, x._2))
    val runFutures: Seq[Future[Unit]] = {
      implicit val ec: ExecutionContextExecutor =
        Utils.newDefaultExecutionContext()
      loaders
        .map(l => Future(l.run(schema)))
        .tap { futures =>
          futures.foreach(_.onComplete {
            case Failure(ex) => throw new DataException(ex)
            case Success(_) =>
              ThreadNameContext.switch(s"transaction[$transactionId]") { _ =>
                logger.info(
                  s"Copy in progress." +
                    s" {done: ${futures.count(_.isCompleted)} / ${futures.size}}"
                )
              }
          })
        }
    }

    runControl

    val cleanupFutures: Seq[Future[Unit]] = {
      implicit val ec: ExecutionContextExecutor =
        Utils.newDefaultExecutionContext()
      loaders.map(l => Future(l.cleanup())).tap { futures =>
        futures.foreach(_.onComplete {
          case Failure(ex) => throw new DataException(ex)
          case _           => // do nothing
        })
      }
    }

    {
      implicit val ec: ExecutionContextExecutor =
        Utils.newDefaultExecutionContext()
      Await.result(Future.sequence(runFutures), Duration.Inf)
      Await.result(Future.sequence(cleanupFutures), Duration("1h"))
    }

    logger.info(s"All of the copies are completed. {done: ${runFutures
      .count(_.isCompleted)} / ${runFutures.size}}")
  }
}
